win32/src/win32evtlog.i

```diff
@@ -824,25 +932,119 @@ static PyObject *PyEvtSeek(PyObject *self, PyObject *args, PyObject *kwargs)
 }
 PyCFunction pfnPyEvtSeek = (PyCFunction) PyEvtSeek;
 
-// @pyswig str|EvtRender|Formats an event into XML text
+// @pyswig o|EvtCreateRenderContext|Creates a render context
+// @comm Accepts keyword args
+static PyObject* PyEvtCreateRenderContext( PyObject *self, PyObject *args, PyObject *kwargs )
+{
+	static char *keywords[]={"Flags", NULL};
+	DWORD flags = EvtRenderContextSystem;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "k:EvtCreateRenderContext", keywords,
+		&flags))	// @pyparm int|Flags||EvtRenderContextSystem or EvtRenderContextUser.  EvtRenderContextValues not currently supported
+		return NULL;
+
+	EVT_HANDLE ret = NULL;
+	Py_BEGIN_ALLOW_THREADS
+	ret = EvtCreateRenderContext( 0, NULL, flags );
+	Py_END_ALLOW_THREADS
+
+	if ( ret == NULL) {
+		return PyWin_SetAPIError("EvtCreateRenderContext", GetLastError() );
+	}
+
+	return PyWinObject_FromEVT_HANDLE(ret);
+}
+PyCFunction pfnPyEvtCreateRenderContext = (PyCFunction) PyEvtCreateRenderContext;
+
+//Returns a list of Event Values in the same order as returned by the EvtRender function
+static PyObject *RenderEventValues( EVT_HANDLE render_context, EVT_HANDLE event )
+{
+	PyObject* ret = NULL;
+	if ( render_context == NULL ) {
+		PyWin_SetAPIError( "EvtRender - Invalid render context for EvtRenderEventValues" );
+		return NULL;
+	}
+
+	DWORD allocated_size = 0;
+	DWORD returned_size = 0;
+	DWORD prop_count = 0;
+	PEVT_VARIANT variants = NULL;
+	BOOL bsuccess = 0;
+	DWORD err = ERROR_SUCCESS;
+
+	Py_BEGIN_ALLOW_THREADS
+	bsuccess = EvtRender( render_context, event, EvtRenderEventValues, allocated_size, variants, &returned_size, &prop_count);
+	Py_END_ALLOW_THREADS
+
+	//bsuccess should always be false here, because we call it initially to get
+	//the size of the buffer
+	if (!bsuccess) {
+		err = GetLastError();
+		if ( err != ERROR_INSUFFICIENT_BUFFER ) {
+			PyWin_SetAPIError( "EvtRender", err );
+			goto cleanup;
+		}
+		else {
+			//allocate buffer size
+			allocated_size = returned_size;
+			variants = (PEVT_VARIANT)malloc( allocated_size );
+			if (variants==NULL){
+				PyErr_NoMemory();
+				goto cleanup;
+			}
+
+			Py_BEGIN_ALLOW_THREADS
+			bsuccess = EvtRender( render_context, event, EvtRenderEventValues, allocated_size, variants, &returned_size, &prop_count);
+			Py_END_ALLOW_THREADS
+		}
+	}
+
+	if ( !bsuccess ) {
+		err = GetLastError();
+		PyWin_SetAPIError( "EvtRender", err );
+		goto cleanup;
+	}
+
+	ret = PyList_New( prop_count );
+	for ( DWORD i = 0; i < prop_count; ++i ) {
+		PyObject* item = PyWinObject_FromEVT_VARIANT( &variants[i] );
```

Comment:
This loop might leave a pending python error, which should be avoided. Is there a reason to not fail entirely here? If so, we should probably `PyErr_Clear()`